Feature/admin invitation model

Geroy4ik opened the pull request:
Задача https://www.notion.so/2a8fee5a455b40edab9e07e6b25309c1?p=c6976cf87c2540fd853f664ceb1b9a0e&pm=s.
Описал модель, сделал ей метод, который возвращает ссылку с токеном. Только я не нашел куда ссылка должна вести хотя бы приблизительно. И сколько ссылка должна жить тоже не ясно, добавил в настройки параметр INVITE_LINK_EXPIRATION_TIME, чтобы можно было настроить.

Geroy4ik committed:
added model AdministratorMailRequest

Geroy4ik committed:
added description to invite_link_expiration_time

Geroy4ik committed:
deleted migration with bad table name

Geroy4ik committed:
added proper migration

Geroy4ik committed:
renamed table name

ADesBiysk commented on src/core/db/models.py:
```diff
@@ -229,3 +234,27 @@ def send_report(self, photo_url: str):
+
+    name = Column(String(100), nullable=False)
+    surname = Column(String(100), nullable=False)
+    email = Column(String(255), nullable=False)
```
Параметры полей нужно будет синхронизировать с этим PR: https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/pull/163

Geroy4ik replied:
Надеюсь email для моей модели не надо делать уникальной? А то еще раз приглашение на эту же почту уже не сделаешь.

ADesBiysk replied:
Под синхронизацией я имел в виду размер полей.
Уникальной точно не нужно. Вероятнее всего, эта таблица будет использоваться и для восстановления пароля. Поэтому часть полей может быть обязательным, но заполняться пустой строкой, в некоторых случаях.

ADesBiysk commented on src/core/db/models.py:
```diff
@@ -229,3 +234,27 @@ def send_report(self, photo_url: str):
+    expired_date = Column(TIMESTAMP, default=get_new_expiration_date, nullable=False)
+
+    def get_link(self):
+        return f"{settings.APPLICATION_URL}/admin/register/{self.token}"
```
С адресом роутера ориентируйся на этот PR: https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/pull/163
Соответственно, у тебя будет GET /invite
Формирование URL лучше убрать из модели в этот метод, поскольку это уже к бизнес-логике относится.

Geroy4ik replied:
Получается в POST запросе: Действующий администратор нажимает в своей панели "пригласить нового члена", заполняет форму и отправляет POST запрос, при котором формируется объект модели AdministratorMailRequest и в ответ получает эту ссылку(или на почту письмо уходит).
Насчет бизнес логики. Это модель приглашения, которая не хранит в себе ссылку, но по требованию может её сформировать, поэтому эта функция, по-моему, здесь уместна.

ADesBiysk replied:
1. Письмо уходит сразу на почту.
2. Мне не нравится, что здесь появляются параметры сервера. Этим мы добавляем новые связи между уровнями (увеличиваем связность), что не желательно.

Geroy4ik committed:
refactored AdministratorMailRequest model,
added router, repository and service

Geroy4ik committed:
remade migration for AdministratorMailRequest

Geroy4ik committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into feature/admin_invitation_model

Geroy4ik committed:
changed router method invite to process_invite,
moved key and token generation to router method

Geroy4ik committed:
fixed key in path from bytes to hex

Geroy4ik commented:
@ADesBiysk Отправку почты тоже мне делать? Просто рассматривал варианты - для FASTAPI вроде как основной [fastapi-mail](https://sabuhish.github.io/fastapi-mail/).  Но тут встала проблема зависимостей: версия 1.2.4 требует stralette версии 0.22, а тот - fastapi 0.88. У нас сейчас fastapi версии 0.82 - я не уверен, что ничего не сломается, если все обновить. Или лучше воспользоваться стандартной библиотекой Python?

Geroy4ik committed:
Merge remote-tracking branch 'origin/develop' into feature/admin_invitation_model

ADesBiysk commented:
> @ADesBiysk Отправку почты тоже мне делать? Просто рассматривал варианты - для FASTAPI вроде как основной [fastapi-mail](https://sabuhish.github.io/fastapi-mail/). Но тут встала проблема зависимостей: версия 1.2.4 требует stralette версии 0.22, а тот - fastapi 0.88. У нас сейчас fastapi версии 0.82 - я не уверен, что ничего не сломается, если все обновить. Или лучше воспользоваться стандартной библиотекой Python?

Да. Почту тоже надо отправлять.
Фреймворк активно развивается, как и наш проект, поэтому лучше перейти на новую версию. Если есть возможность, то обнови до новой версии и попробуй несколько разных методов.

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -0,0 +1,41 @@
+        # TODO Добавить отправку сообщения на электронную почту
+        return AdministratorMailRequestResponse(url=request.url_for('process_invite', key=key.hex()))
+
+    @router.get(
```
Этот роутер не нужен. В другой задаче сразу будет метод регистрации по ссылке.

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -0,0 +1,41 @@
+        self, request: Request, invitation_data: AdministratorMailRequestRequest
+    ) -> AdministratorMailRequestResponse:
+        key = secrets.token_bytes()
+        token = hashlib.sha256(key).hexdigest()
```
Это надо перенести в service. Сейчас создание и проверка токена расположены в разных слоях, что совсем плохо.

ADesBiysk commented on src/core/services/administrator_mail_request.py:
```diff
@@ -0,0 +1,29 @@
+        new_invite = AdministratorMailRequest(**invitation_data.dict(), token=token)
+        return await self.__administrator_mail_request_repository.create(new_invite)
+
+    async def get_invite_by_token(self, token: str) -> Optional[AdministratorMailRequest]:
```
Зачем нужен этот прокси-метод?

ADesBiysk commented on src/core/services/administrator_mail_request.py:
```diff
@@ -0,0 +1,29 @@
+    async def get_invite_by_token(self, token: str) -> Optional[AdministratorMailRequest]:
+        return await self.__administrator_mail_request_repository.get_invite_by_token(token)
+
+    async def verify_token(self, token: str) -> AdministratorMailRequest:
```
Этот метод будет использоваться при регистрации, и назвать его лучше get_active_mail_request.
В этой задаче он не будет использоваться.

Geroy4ik committed:
moved key generation to service,
renamed methods, removed redundant method,
moved expiration date check to repository statement

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -0,0 +1,41 @@
+        self, request: Request, invitation_data: AdministratorMailRequestRequest
+    ) -> AdministratorMailRequestResponse:
+        key = secrets.token_bytes()
+        token = hashlib.sha256(key).hexdigest()
```
Такая генерация токена - перебор. Здесь достаточно просто uuid4. Мы его будет просто проверять по базе. 
Если это защита от случая, что злоумышленник получил доступ к БД и не воспользовался токеном, то это не нужно. Если у злоумышленника есть доступ к БД, то он может получить доступ к системе намного проще.
А если перехватит письмо, то этот способ не поможет.

Geroy4ik replied:
А нужно ли нам тогда поле token вообще? С таким же успехом можно id отправлять.

ADesBiysk replied:
Это разные поля и выполняют они разные функции. Что будет, если в будущем потребуется изменить алгоритм генерации токена, например, сделать его удобным для ручного ввода (утрированно, приглашение будет распечатано на бумаге). Это потребует изменения схемы БД, а в данном случае нет.

Geroy4ik committed:
updated fastapi, installed fastapi-mail,
added email sending

Geroy4ik committed:
deleted obsolete migration

Geroy4ik committed:
Merge branch 'feature/admin_invitation_model_experimental' into feature/admin_invitation_model

Geroy4ik committed:
updated .env.example with mail settings

Geroy4ik commented:
Добавил отправку email, но как быть с настройками?

ADesBiysk commented on requirements.txt:
```diff
@@ -1,14 +1,66 @@
-python-telegram-bot==20.0a4
-pytz==2022.6
-uvicorn==0.18.3
+aiohttp==3.8.3 ; python_version >= "3.10" and python_version < "4.0"
```
Зачем здесь такая "портянка". Нужно только добавить основные библиотеки. Зависимости сами поставятся. Мы сами формируем этот файл, и если будут проблемы, то можно будет сделать freeze, но не сейчас.
Зачем это: ; python_version >= "3.10" and python_version < "4.0"

Geroy4ik replied:
Это через poetry такой файл получился.

ADesBiysk commented:
> Добавил отправку email, но как быть с настройками?

Проще всего сделать так. Зарегистрируй любую почту. Добавь в worksflow запись в .env основных настроек из секретов github. А секреты вышли мне в личку и я их добавлю в github.

Geroy4ik committed:
remade requirements.txt, added mail secrets to
deploy_stage workflow

Geroy4ik commented:
Кстати, тут давненько check_migrations валятся на старте и получается всего 2 проверки вместо 3.

AntonZelinsky commented on .env.example:
```diff
@@ -8,3 +8,9 @@ BOT_WEBHOOK_MODE=False # запустить бота в режиме webhook(tru
 DEBUG=False # Включение (True) | Выключение (False) режима отладки
+MAIL_SERVER=smtp.mail.ru # адрес постового сервиса
+MAIL_PORT=465 # порт для подключения к почтовому сервису
+MAIL_LOGIN= # логин для подключения к почтовому сервису
+MAIL_PASSWORD= # пароль для подключения к почтовому сервису
```
Захардкодь тут какой-нибудь тестовый логи/пароль, чтобы запуская локально у всех работал функционал отправки почты.

AntonZelinsky commented on .github/workflows/deploy_stage.yaml:
```diff
@@ -89,6 +89,11 @@ jobs:
+            echo MAIL_LOGIN=${{ secrets.MAIL_LOGIN }} >> .env
+            echo MAIL_PASSWORD=${{ secrets.MAIL_PASSWORD }} >> .env
+            echo MAIL_STARTTLS=${{ secrets.MAIL_STARTTLS }} >> .env
+            echo MAIL_SSL_TLS=${{ secrets.MAIL_SSL_TLS }} >> .env
```
Предлагаю, не выносить сюда все эти настройки протокола. пусть они остаются в сетингах фастапи, но тут будет только логин/пароль. КОгда будем катиться на прод подставим другие значения или придумаем как можно перерабоатть. Сейчас не усложняем.

AntonZelinsky commented on src/api/routers/administrator.py:
```diff
@@ -0,0 +1,36 @@
+        try:
+            await Email().send_invitation_link([invite.email], url, invite.name)
+        except Exception as exc:
+            raise EmailSendException(invite.email, invite.id, exc)
```
Выглядит как бизнес логика, значит, в сервис нужно отправить.

AntonZelinsky replied:
И исключение перехватывать и возвращать своё в емаил провайдере, а не в сервисе.(

AntonZelinsky commented on src/api/routers/report.py:
```diff
@@ -50,23 +50,27 @@ async def approve_task_status(
+        "/{report_id}/decline",
+        status_code=HTTPStatus.OK,
+        summary="Отклонить задание.",
+    )
```
Почему тут перенос на новую строку? Выглядит так, что у тебя как-то не правильно настроен автоформатер или иде меняет форматирование. Не нужно менять то, что ты не затрагиваешь, а форматирование должно быть как у всех. Длинна на 120 символов, если я не ошибаюсь. Посмотри, в твоём ПРе много где поменялось форматирование, исправь везде назад.

Geroy4ik replied:
Да, это fast-mail установила black в проект, а он подцепил мои внешние настройки на ширину 79.

AntonZelinsky commented on src/api/routers/report.py:
```diff
@@ -50,23 +50,27 @@ async def approve_task_status(
         report_id: UUID,
         request: Request,
-    ) -> HTTPStatus.OK:
+    ) -> None:
```
И почему ты поудалял много где анотацию, сообщающую, что вернётся код 200?

Geroy4ik replied:
В последних версиях fastapi pydantic  [валидирует тип возврата по аннотации](https://fastapi.tiangolo.com/tutorial/response-model/#invalid-return-type-annotations). Тогда лучше поставлю параметр `response_model=None`, чтобы он игнорировал возвращаемый тип.

AntonZelinsky commented on src/core/settings.py:
```diff
@@ -71,3 +83,8 @@ def get_settings():
+    "days": 1,
+    "hours": 0,
+    "minutes": 0,
+}  # время существования ссылки для приглашения на регистрацию
```
Зачем так сложно? почему бы не взять просто dateime?

AntonZelinsky commented on src/core/services/administrator_mail_request.py:
```diff
@@ -0,0 +1,40 @@
+from src.core.exceptions import AdministratorMailRequestInvalid
+
+
+class AdministratorMailRequestService:
```
Плохое название. Если завтра нужно будет отправлять приглашение в телеграм, ты будешь делать ещё один сервис и модели или пере используешь? 
 Лучше назвать AdministratorInvitetion. И создать дополнительную папку admin или администратор и уже туда ложить это. Чтобы отделить логику основного приложения, от административной части

AntonZelinsky commented on src/core/services/administrator_mail_request.py:
```diff
@@ -0,0 +1,40 @@
+            AdministratorMailRequest(**invitation_data.dict())
+        )
+
+    async def get_active_mail_request(self, token: str) -> AdministratorMailRequest:
```
Кажется, что это нигде не используется

AntonZelinsky commented on src/core/services/administrator_mail_request.py:
```diff
@@ -0,0 +1,40 @@
+        mail_request = await self.__administrator_mail_request_repository.get_mail_request_by_token(token)
+        if not mail_request:
+            raise AdministratorMailRequestInvalid
+        return mail_request
```
Какой смысл делать тут отрицание? только сложнее ведь становится

AntonZelinsky replied:
А вообще, предлагаю поступить тут как в методе get репозитория. Всегда возвращать значение либо кидать исключение. Этим нет смысла заниматься сервису.

AntonZelinsky commented on src/core/db/models.py:
```diff
@@ -229,3 +263,21 @@ def send_report(self, photo_url: str):
         self.number_attempt += 1
+
+
+class AdministratorMailRequest(Base):
```
Очень плохое название модели, писал в других местах. + сюда же будет писаться токен для восстановления пароля, можно сразу прикинуть общее название, если не получится, то называть как сервис, а потом переименуем.

AntonZelinsky commented on src/core/db/models.py:
```diff
@@ -229,3 +263,21 @@ def send_report(self, photo_url: str):
+    __tablename__ = "administrator_mail_requests"
+
+    def get_new_expiration_date():
+        return datetime.utcnow() + timedelta(**INVITE_LINK_EXPIRATION_TIME)
```
метод должен быть статическим, если не используются другие члены этого класса. Но вообще, этому тут не место, он должен быть приватным в сервисе. и так INVITE_LINK_EXPIRATION_TIME никогда не импортируй! не понятно ни что это ни откуда.

AntonZelinsky commented on src/core/email.py:
```diff
@@ -0,0 +1,63 @@
+from src.core.settings import ORGANIZATIONS_EMAIL, settings
+
+
+class Email:
```
Назови EmailProvider

AntonZelinsky commented on src/core/email.py:
```diff
@@ -0,0 +1,63 @@
+class Email:
+    """Класс для отправки электронных писем."""
+
+    async def send_mail(
```
приватной должна быть

Geroy4ik committed:
review fixes, hardcoded test mail credentials

Geroy4ik committed:
renamed AdministratorMailRequest to
AdministratorInvitation

Geroy4ik committed:
temporary renamed router

Geroy4ik committed:
deleted duplicate file

Geroy4ik committed:
temp poetry.lock from develop branch

Geroy4ik committed:
Merge branch 'develop' into feature/admin_invitation_model

Geroy4ik committed:
new poetry.lock after merging conflicts

Geroy4ik committed:
renamed exception

Geroy4ik committed:
renamed  Administrator to AdministratorInvitation

Geroy4ik committed:
remade migration file
